cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/foreign_assets.rs

```diff
@@ -0,0 +1,430 @@
+
+	assert!(ah_sender_balance_after < ah_sender_balance_before);
+	assert!(penpal_receiver_balance_after > penpal_receiver_balance_before);
+}
+
+// ==============================================================================================
+// ==== Bidirectional Transfer - Reserve-based Foreign Asset - Penpal<->AssetHub ====
+// ==============================================================================================
+/// Transfers of foreign asset from Penpal to AssetHub and back. Foreign Asset is not registered
+/// with Asset Hub as a trusted reserve, ergo teleports are not available and reserve-transfers are
+/// to be used. Also verifies that teleporting the asset fails both ways.
+#[test]
+fn bidirectional_reserve_transfer_foreign_asset_between_penpal_and_asset_hub() {
+	let sender = PenpalASender::get();
+	let receiver = AssetHubWestendReceiver::get();
+	let new_asset_id = 42;
+	let asset_amount_to_send = ASSET_HUB_WESTEND_ED * 10_000;
+	let (asset_location_on_penpal, foreign_asset_location_on_ah) =
+		set_up_foreign_asset(sender.clone(), new_asset_id, asset_amount_to_send, false);
+
+	////////////////////////////////////////
+	// Reserve-transfer it from Penpal to AH
+	////////////////////////////////////////
+
+	let penpal_sender_balance_before = assets_balance_on!(PenpalA, new_asset_id, &sender);
+	let ah_receiver_balance_before =
+		foreign_balance_on!(AssetHubWestend, foreign_asset_location_on_ah.clone(), &receiver);
+
+	let dest = PenpalA::sibling_location_of(AssetHubWestend::para_id());
+	let assets: Assets =
+		vec![(asset_location_on_penpal.clone(), asset_amount_to_send).into()].into();
+	// execute xcm from penpal to asset hub
+	PenpalA::execute_with(|| {
+		// xcm to be executed at dest
+		let xcm_on_dest = Xcm(vec![
+			// since this is the last hop, we don't need to further use any assets previously
+			// reserved for fees (there are no further hops to cover delivery fees for); we
+			// RefundSurplus to get back any unspent fees
+			RefundSurplus,
+			DepositAsset { assets: Wild(All), beneficiary: receiver.clone().into() },
+		]);
+		// teleporting the asset fails
+		let xcm = Xcm::<()>(vec![
+			WithdrawAsset(assets.clone().into()),
+			SetFeesMode { jit_withdraw: true },
+			InitiateTransfer {
+				destination: dest.clone(),
+				remote_fees: Some(AssetTransferFilter::Teleport(assets.clone().into())),
+				preserve_origin: false,
+				assets: BoundedVec::new(),
+				remote_xcm: xcm_on_dest.clone(),
+			},
+		]);
+		assert!(matches!(
+			<PenpalA as PenpalAPallet>::PolkadotXcm::execute(
+				<PenpalA as Chain>::RuntimeOrigin::signed(sender.clone()),
+				bx!(xcm::VersionedXcm::from(xcm.into())),
+				Weight::MAX,
+			),
+			Err(sp_runtime::DispatchErrorWithPostInfo { .. }),
+		));
+		// reserve-transferring the asset works
+		let xcm = Xcm::<()>(vec![
+			WithdrawAsset(assets.clone().into()),
+			SetFeesMode { jit_withdraw: true },
+			InitiateTransfer {
+				destination: dest,
```

Comment:
We probably want to write examples where most assets are sent via `assets`, right? Instead of `remote_fees`